src/server/scene/surface_stack.h

```diff
@@ -82,43 +83,57 @@ class SurfaceStack : public compositor::Scene, public input::Scene, public shell
 
     virtual void remove_surface(std::weak_ptr<Surface> const& surface) override;
 
+    void raise(Surface const* surface);
     virtual void raise(std::weak_ptr<Surface> const& surface) override;
-
     void raise(SurfaceSet const& surfaces) override;
 
     void add_surface(
         std::shared_ptr<Surface> const& surface,
         input::InputReceptionMode input_mode) override;
-    
+
     auto surface_at(geometry::Point) const -> std::shared_ptr<Surface> override;
 
     void add_observer(std::shared_ptr<Observer> const& observer) override;
     void remove_observer(std::weak_ptr<Observer> const& observer) override;
-    
+
     // Intended for input overlays, as described in mir::input::Scene documentation.
     void add_input_visualization(std::shared_ptr<graphics::Renderable> const& overlay) override;
     void remove_input_visualization(std::weak_ptr<graphics::Renderable> const& overlay) override;
-    
+
     void emit_scene_changed() override;
 
 private:
     SurfaceStack(const SurfaceStack&) = delete;
     SurfaceStack& operator=(const SurfaceStack&) = delete;
     void create_rendering_tracker_for(std::shared_ptr<Surface> const&);
     void update_rendering_tracker_compositors();
+    void insert_surface_at_top_of_depth_layer(std::shared_ptr<Surface> const& surface);
+
+    /**
+     * Converts a MirDepthLayer into an index for a list of layers (0 is bottom-most layer)
+     */
+    static auto mir_depth_layer_to_index(MirDepthLayer depth_layer) -> unsigned int;
```

Comment:
I believe you could mark this as constexpr?

Reply:
Not sure why, as we never use it as `constexpr`, but ok

Reply:
Never mind. Due to a GCC bug (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=67371), there isn't an easy way to raise an error if we fall through the switch. Since it's a private function and we don't use it `constexpr`, I'm not going to bother working around that.